File: C++/Tower_of_Hanoi.cpp
// Tower of Hanoi

// We need to move n disks from source to destination

// Rules :-
// 1. Only one disk can be moved in 1 step
// 2. Smaller disks are always kept on top of larger disks

// Algorithm :-
// 1. Move n-1 disks from source to helper
// 2. Move 1 disk from source to destination
// 3. Move n-1 disks from helper to destination

#include<iostream>
#include<vector>

using namespace std;

void TowerOfHanoi(int n, string source, string helper, string destination){
    if (n == 1){
        cout << "Move disk 1 from " << source << " to " << destination << endl;
        return;
    }

    TowerOfHanoi(n - 1, source, destination, helper);
    cout << "Move disk " << n << " from " << source << " to " << destination << endl;
    TowerOfHanoi(n - 1, helper, source, destination);
}

int main(){
    int n;
    cin >> n;
    TowerOfHanoi(n, "A", "B", "C");
    return 0;
}

File: Java/Tower_Of_Hanoi.java
public class Tower_Of_Hanoi {
    public static void Tower_of_Hanoi(int n,String src,String helper,String dest){
        if(n==0){
            System.out.println("Transfer Disk : " + n + " From " + src + " -> " + dest);
            return;
        }
        Tower_of_Hanoi(n-1,src,dest,helper);
        System.out.println("Transfer Disk : " + n + " From " + src + " -> " + dest);
        Tower_of_Hanoi(n-1,helper,src,dest);
    }
    public static void main(String[] args){
        int n=3;

        Tower_of_Hanoi(n,"A","B","C");
    }
}
